source/common/config/new_grpc_mux_impl.cc

```diff
@@ -153,13 +155,28 @@ GrpcMuxWatchPtr NewGrpcMuxImpl::addWatch(const std::string& type_url,
 // subscription will enqueue and attempt to send an appropriate discovery request.
 void NewGrpcMuxImpl::updateWatch(const std::string& type_url, Watch* watch,
                                  const std::set<std::string>& resources,
-                                 bool creating_namespace_watch) {
+                                 const bool creating_namespace_watch) {
   ASSERT(watch != nullptr);
   auto sub = subscriptions_.find(type_url);
   RELEASE_ASSERT(sub != subscriptions_.end(),
                  fmt::format("Watch of {} has no subscription to update.", type_url));
-  auto added_removed = sub->second->watch_map_.updateWatchInterest(watch, resources);
-  if (creating_namespace_watch) {
+  // If this is a glob collection subscription, we need to compute actual context parameters.
+  std::set<std::string> xdstp_resources;
+  if (resources.size() == 1 && XdsResourceIdentifier::hasXdsTpScheme(*resources.begin())) {
```

Comment:
This seems like we are adding another responsibility to add/update watch code. Perhaps this can be extracted into a dedicated class? Will be easier to test too.

Reply:
It's a good question, how do you envisage this looking? The key thing to keep in mind is that we need some code that will responsible for mapping from a combination of resource name, node-specific context parameters and dynamic context parameters, to low-level resource names on subscription. This can happen outside of the mux/watch code. We also need code inside the mux/watch code to be able to implement `xdstp://` semantics, e.g. to be able to tell "I've received a resource, which really belongs to this glob collection, since the resource name modulo * matches".